### PR TITLE
Feature cluster variable for route. 

### PR DESCRIPTION
**Pull request by wangfakang**
### Issues associated with this PR

Your PR should present related issues you want to solve.

### Solutions
You should show your solutions about the issues in your PR, including the overall solutions, details and the changes. At this time, Chinese is allowed to describe these.

### UT result
Unit Test is needed if the code is changed, your unit test should cover boundary cases, corner cases, and some exceptional cases. And you need to show the UT result.

### Benchmark
If your code involves the processing of every request, you should give the Benchmark Result.

### Code Style
+ Make sure `Goimports` has run
+ Show `Golint` result

### CODE REVIEW

File: pkg/router/base_rule.go [hunk 1]
```diff
@@ -206,10 +206,16 @@ func (rri *RouteRuleImplBase) RedirectRule() api.RedirectRule {
 // Select Cluster for Routing
 // if weighted cluster is nil, return clusterName directly, else
 // select cluster from weighted-clusters
-func (rri *RouteRuleImplBase) ClusterName() string {
+func (rri *RouteRuleImplBase) ClusterName(ctx context.Context) string {
 	if len(rri.weightedClusters) == 0 {
-		return rri.defaultCluster.clusterName
+		// if both 'cluster_name' and 'cluster_variable' are configured, 'cluster_name' is preferred.
```
Comment (nejisama) [inline pkg/router/base_rule.go:211]: 这个感觉是不是不应该这样。 cluster_name 是配置路由要选择哪个cluster，而变量这个需求的目的看上去是希望通过变量来修改路由的结果。 那么这里是不是可以考虑，将cluster_name 的结果保存在变量里，在选择Cluster的时候直接从变量拿？

Comment (nejisama) [reply pkg/router/base_rule.go:211]: 这样变量的机制还可以对WeightCluster 也可以生效，没有必要再单独搞一个配置什么的

File: pkg/router/base_rule.go [hunk 2]
```diff
@@ -206,10 +206,21 @@ func (rri *RouteRuleImplBase) RedirectRule() api.RedirectRule {
...
 // if weighted cluster is nil, return clusterName directly, else
 // select cluster from weighted-clusters
-func (rri *RouteRuleImplBase) ClusterName() string {
+func (rri *RouteRuleImplBase) ClusterName(ctx context.Context) string {
 	if len(rri.weightedClusters) == 0 {
-		return rri.defaultCluster.clusterName
+		// If both 'cluster_name' and 'cluster_variable' are configured, 'cluster_name' is preferred.
+		if rri.defaultCluster.clusterName != "" {
+			return rri.defaultCluster.clusterName
+		} else if clusterName, err := variable.GetVariableValue(ctx, rri.routerAction.ClusterVariable); err == nil {
```
Comment (taoyuanyuan) [inline pkg/router/base_rule.go:214]: else if -> if

Comment (taoyuanyuan) [reply pkg/router/base_rule.go:214]: if err != nil , should return "error string" or "" ?

Comment (wangfakang) [reply pkg/router/base_rule.go:214]: > if err != nil , should return "error string" or "" ?

If an error occurred while get cluster name from variable, then we need to get it from `DefaultCluster`.

Comment (wangfakang) [reply pkg/router/base_rule.go]: > else if -> if

done.

**Review by taoyuanyuan — DISMISSED** [review]
LGTM

**Review by taoyuanyuan — APPROVED** [review]
lgtm

**Review by nejisama — APPROVED** [review]

**Pull request merged**